### PR TITLE
Add the response timeout corresponding to each request that is made …

### PR DESCRIPTION
**Pull request by minwoox**
…in `RetryingClient`

Motivation:

`ResponseTimeout` was applied to the whole requests in `RetryingClient`. A client could not
handle the responses with the timeout individually.

Modification:

- Add `responseTimeoutMillisForEachAttempt` when `RetryingClient` is made
- Remove `shouldRetry(request, exception)` in `RetryStrategy`

Result:

- A client can handle each response with the timeout

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java [hunk 1]
```diff
@@ -54,11 +52,23 @@
     }
 
     /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<RpcRequest, RpcResponse>, RetryingRpcClient>
+    newDecorator(RetryStrategy<RpcRequest, RpcResponse> retryStrategy,
+                 int defaultMaxAttempts, long responseTimeoutMillisOfEachReq) {
```
Comment (imasahiro) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java]: Could you add javadoc for `responseTimeoutMillisOfEachReq`?
`@param responseTimeoutMillisOfEachReq response timeout for each request. {@code 0} disables the timeout.`

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java]: That's good idea~! I will add it.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 2]
```diff
@@ -133,43 +136,71 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, lon
...
+            resetDeadline();
         }
 
-        void init(long responseTimeoutMillis) {
-            setDeadlineOfThisRequest(responseTimeoutMillis);
-            lastBackoff = null;
-            totalAttemptNo = 1;
+        void resetDeadline() {
+            deadlineNanos = responseTimeoutMillis <= 0 ?
+                            0 : System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(responseTimeoutMillis);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: It is possible for `deadlineNanos` to be `0` even if `responseTimeoutMillis` is greater than 0.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 3]
```diff
@@ -133,43 +136,71 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, lon
...
+
+        /**
+         * Returns the smaller value between {@code responseTimeoutMillisOfEachReq} and
+         * remaining {@code responseTimeoutMillis}.
+         *
+         * @return 0 if the response timeout for both of each request and whole retry is disabled or
+         *         -1 if the elapsed time from the first request has passed {@code responseTimeoutMillis}
+         */
+        long responseTimeoutMillis() {
+            if (deadlineNanos == 0) { // response timeout to the whole is disabled.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: `responseTimeoutMillis <= 0` ?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 4]
```diff
@@ -133,43 +136,71 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, lon
...
+                return responseTimeoutMillisOfEachReq;
+            }
+
+            final long responseTimeoutMillis = TimeUnit.NANOSECONDS.toMillis(deadlineNanos - System.nanoTime());
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: How about `actualResponseTimeoutMillis`?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 5]
```diff
@@ -133,43 +136,71 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, lon
...
+
+            if (responseTimeoutMillis <= 0) { // 0 is not from the first, but subtracted to that, it's timeout.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: I'm not sure I understood the comment. Could you rephrase it a little bit?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java [hunk 6]
```diff
@@ -72,6 +73,21 @@ public T defaultMaxAttempts(int defaultMaxAttempts) {
...
+     * Sets the {@code responseTimeoutMillisOfEachReq}. When requests in {@link RetryingClient} are made,
+     * corresponding responses are timed out by this value.
+     * It will be set by the default value in {@link Flags#defaultResponseTimeoutMillis()}, if the client
+     * dose not specify.
+     *
+     * @return {@link T} to support method chaining.
+     */
+    public T responseTimeoutMillisOfEachReq(long responseTimeoutMillisOfEachReq) {
+        checkArgument(responseTimeoutMillisOfEachReq >= 0,
+                      "responseTimeoutMillisOfEachReq: %s (expected: >=0)", responseTimeoutMillisOfEachReq);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java]: `>=0` -> `>= 0` (space between `>=` and `0`)

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java [hunk 7]
```diff
@@ -72,6 +73,21 @@ public T defaultMaxAttempts(int defaultMaxAttempts) {
...
 
     /**
+     * Sets the {@code responseTimeoutMillisOfEachReq}. When requests in {@link RetryingClient} are made,
+     * corresponding responses are timed out by this value.
+     * It will be set by the default value in {@link Flags#defaultResponseTimeoutMillis()}, if the client
+     * dose not specify.
+     *
+     * @return {@link T} to support method chaining.
+     */
+    public T responseTimeoutMillisOfEachReq(long responseTimeoutMillisOfEachReq) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java]: What about `responseTimeoutMillisForEachAttempt`?

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java]: That's a good name. I'v been trying to figure out a better name, but I couldn't.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java (hunk 3 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: Fixed :-)

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java (hunk 4 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: Fixed :-)

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java (hunk 6 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java]: Fixed :-)

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 8]
```diff
@@ -133,43 +136,71 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, lon
...
-            totalAttemptNo = 1;
+        void resetDeadline() {
+            deadlineNanos = responseTimeoutMillis <= 0 ?
+                            0 : System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(responseTimeoutMillis);
+            responseTimeoutOccurred = false;
+        }
+
+        /**
+         * Returns the smaller value between {@code responseTimeoutMillisOfEachReq} and
+         * remaining {@code responseTimeoutMillis}.
```
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: Fixed :-)

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java (hunk 2 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: Got it. Thank you~!!

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java (hunk 5 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: Rephrased!

**Comment by hyangtack** [comment]
The feature and specification are okay for me. 👍 There are some codes which are unnecessarily called if timeout occurred. (in `RetryingHttpClient#retry`) I think it would be better to fix them and to rearrange methods for readability. :-)

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java [hunk 9]
```diff
@@ -93,53 +113,64 @@
...
+                           if (backoffOpt.isPresent()) {
+                               final long millisAfter =
+                                       useRetryAfter ? getRetryAfterMillis(resDuplicator.duplicateStream())
+                                                     : -1;
+                               resDuplicator.close();
+
+                               long nextDelay;
+                               try {
+                                   nextDelay = getNextDelay(ctx, backoffOpt.get(), millisAfter);
+                                   if (nextDelay < 0) { // exceed the number of max attempt
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java]: How about throwing `RetryGiveUpException` in `getNextDelay` instead of returning `-1` so that we can remove this `if` block?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryStrategy.java [hunk 10]
```diff
@@ -89,20 +85,10 @@
     /**
      * Returns {@link CompletableFuture} with {@link Optional} that contains {@link Backoff} which will be
      * used for retry. If the condition does not match, this needs to return {@link Optional#EMPTY}
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryStrategy.java]: Not directly related with this PR, but isn't it `{@link Optional#empty()}`?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 11]
```diff
@@ -123,53 +122,84 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, lon
         requireNonNull(backoff, "backoff");
         final State state = ctx.attr(STATE).get();
         final int currentAttemptNo = state.currentAttemptNoWith(backoff);
+
         if (currentAttemptNo < 0) {
+            // exceed the default number of max attempt.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: exceed -> Exceeded?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java [hunk 12]
```diff
@@ -123,53 +122,84 @@ protected final long getNextDelay(ClientRequestContext ctx, Backoff backoff, lon
...
             return -1;
         }
 
         long nextDelay = backoff.nextDelayMillis(currentAttemptNo);
         if (nextDelay < 0) {
+            // exceed the number of max attempt in the backoff.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: exceed -> Exceeded?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java (hunk 9 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java]: That's a great idea~!! I will change it.

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java [hunk 13]
```diff
@@ -72,6 +73,22 @@ public T defaultMaxAttempts(int defaultMaxAttempts) {
...
 
     /**
+     * Sets the {@code responseTimeoutMillisForEachAttempt}. When requests in {@link RetryingClient} are made,
+     * corresponding responses are timed out by this value. {@code 0} disables the timeout.
+     * It will be set by the default value in {@link Flags#defaultResponseTimeoutMillis()}, if the client
+     * dose not specify.
+     *
+     * @return {@link T} to support method chaining.
+     */
+    public T responseTimeoutMillisForEachAttempt(long responseTimeoutMillisForEachAttempt) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java]: How about adding `responseTimeoutForEachAttempt(Duration)` as well, which delegates to this method?

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java [hunk 14]
```diff
@@ -54,70 +57,82 @@
...
+        }
+        final RpcResponse response = executeDelegate(ctx, req);
+
         retryStrategy().shouldRetry(req, response).handle(voidFunction((backoffOpt, unused) -> {
             if (backoffOpt.isPresent()) {
-                retry0(ctx, backoffOpt.get(), req, action, responseFuture);
+                long nextDelay;
+                try {
+                    nextDelay = getNextDelay(ctx, backoffOpt.get());
+                    if (nextDelay < 0) { // exceed the number of max attempt
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java]: exceed -> Exceeded

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java [hunk 15]
```diff
@@ -54,70 +57,82 @@
...
+                        return;
+                    }
+                } catch (ResponseTimeoutException e) {
+                    responseFuture.completeExceptionally(e);
+                    return;
+                }
+
+                EventLoop eventLoop = ctx.contextAwareEventLoop();
+                if (nextDelay == 0) {
+                    eventLoop.submit(() -> doExecute0(ctx, req, responseFuture));
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java]: If you are not interested in the returned `Future`, you can just use `execute()`

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java [hunk 16]
```diff
@@ -93,53 +113,64 @@
...
+                                       return;
+                                   }
+                               } catch (ResponseTimeoutException e) {
+                                   closeOnException(deferredRes, rootReqDuplicator, e);
+                                   return;
+                               }
+
+                               final EventLoop eventLoop = ctx.contextAwareEventLoop();
+                               if (nextDelay == 0) {
+                                   eventLoop.submit(() -> doExecute0(ctx, rootReqDuplicator, deferredRes));
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/retry/RetryingHttpClient.java]: `execute(() -> ...)`

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java (hunk 11 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: Fixed. Thank you~!

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java (hunk 12 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java]: Fixed. Thank you~!

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java (hunk 14 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java]: Fixed. Thank you~!

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java (hunk 15 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java]: Just copied it from the previous codes without thinking :-) Thank you~!

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java (hunk 13 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryingClientBuilder.java]: Added~!! Thank you~!

File: core/src/main/java/com/linecorp/armeria/client/retry/RetryStrategy.java (hunk 10 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/retry/RetryStrategy.java]: Oops~!! Thank you~!

**Comment by imasahiro** [comment]
Thank you @minwoox ~
